// @nucypher/taco/dist/cjs/conditions/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.predefined = exports.base = exports.context = exports.ConditionFactory = exports.conditionExpr = exports.condition = exports.compound = void 0;
const base = __importStar(require("./base"));
exports.base = base;
const predefined = __importStar(require("./predefined"));
exports.predefined = predefined;
exports.compound = __importStar(require("./compound-condition"));
exports.condition = __importStar(require("./condition"));
exports.conditionExpr = __importStar(require("./condition-expr"));
var condition_factory_1 = require("./condition-factory");
Object.defineProperty(exports, "ConditionFactory", { enumerable: true, get: function () { return condition_factory_1.ConditionFactory; } });
exports.context = __importStar(require("./context"));
